### DISCUSSION TITLE
Images Revert on Products Routinely and Disappear

### DISCUSSION THREAD

**Discussion by osseonews** [swellstores/community#141] [Help] [answered]
Just thought I'd post this constant bug with images in the hope that someone else experienced this, b/c even if this happens all the time with our products, I can't get Swell to actually address this extremely serious bug in the platform. So maybe if someone else actually has seen this, it will be acknowledged and a fix deployed.

The bug is simple: Out of nowhere the images for products simply revert to an earlier version of the images and we lose newer prior images. Initially, when this started to happen, I thought I was going nuts and maybe we forgot to actually upload new images. But after this happens every single week for months, you realize it's not you, but there is a serious bug. Didn't we upload new images for this product a few weeks ago and weren't they on the product ? Why are the images no longer there? Why are the images now the same images from months ago before the last update? What happened to the new images which we saved?

The only solution to this problem is to upload all the images again to the product and resave the product and pray that they don't disappear again in a few weeks. Needless this to say having to reupload images to the same product every few weeks is not fun, and it hurts sales. It's impossible to go thru every product every day to ensure that the images have not mysteriously been deleted.

Of course, when we report this problem, we are asked to replicate it. But, how can we replicate it when our images randomly disappear from the product and revert to older images? How do we know when this will happen or why? Are we supposed to take screenshots of when we upload new images to prove that we did it correctly and then show how all the images disappear randomly?

IMHO, there seems to be some cache employed for the image data and product data in a store and this cache is entirely corrupted. The bug might be related to a similar bug which asks to resave a product, even if we didn't make any changes whatsoever to the product. But, I honestly don't know. All I know is that the product image data on Swell randomly disappears and it has nothing to do with anything we do, as we are so scared to touch anything in the dashboard anymore for fear of some random change.

The only solution seems to be to move all our product data to a stable CMS, which is a major pain, as part of the appeal of Swell is having a good product CMS. Luckily, we have been backing up images, so we can recover the lost images on Swell for the time being. but, it would be nice to not have to worry every single day that the images we uploaded for products that help to sell our product don't randomly disappear.

**Comment by ericingram** [comment]
I checked with the team and we haven't had anyone else report this issue, which suggests it could be caused by something on your frontend. How are you fetching products and could they be getting cached somewhere else?

Another approach could be to look at the events on a specific product where you saw this behavior and see whether the image URLs changed at all. If they did, there will be a log of the change in our system and we could track it down.

Image URLs have the MD5 hash of the contents in the URL itself, which means the URL has to change in order for the image to change.

**Reply by osseonews** [reply]
Thanks but the images disappear from the Swell dashboard and are not cached anywhere. It doesn't have to do with our frontend,  as how would our frontend delete images form Swell's backend? We don't call any sort of function like that at all. We only upload images from the Swell dashboard for the product, and save the product in swell. Then a few weeks later the images disappear and the images for the product revert back to an older state.  Incidentally, this happens many times in instances unrelated to images. So you can go to a product, not change a thing, and then get an alert that something as changed and to save the product, even if you haven't changed a thing. I assume the 2 issues are related in some fashion. There is a bug in your dashboard somewhere and/or some caching mechanism on your backend. I'm sure someone else will report this with time. I guess, for now, we will just source the images for our products from another backend, as this is becoming a major headache to manage.

**Reply by ericingram** [reply]
We did some quick digging on one of the products you shared with the support team and found that it's being caused by a cart update where old product data is being passed into the item object. This causes the linked product to get updated.

Link fields like `items.product` are often used to expand linked records, but it also worked in reverse for updating those linked records.

The request causing this looks like:

```
PUT /carts/62c38f97edeb7b00139b694f
{
  "items": [
    {
      "product": { "images": [ OLD_IMAGES ] }
    }
  ]
}
```

You can solve this by removing the `product` object from your cart update.

**Reply by osseonews** [reply]
We don't have any requests anywhere that update product images like that. We do not put "images" in anywhere in our code. You can add or subtract items from the cart on our website and you will never see that in the request headers. I don't know where that request comes from.

**Reply by osseonews** [reply]
Our cart mutations are based on NextJS commerce repo. just copied from there exactly. I don't see any puts there either which send in image for products when updating the cart.

**Reply by osseonews** [reply]
The only thing we ever do with images is a GET request, which is just part of getting product data. We never mutate images in any API call anywhere, nor does NextJS repo from what I can tell. We also never get any array of old images in our data, and so would have no ability to overwrite images, b/c we don't have the old image data anywhere. We just pull the image data from the Swell API when we get a product.

**Reply by osseonews** [reply]
Look thru our code again, and the only other place we put to the cart in our code, other than regular cart functionality, is when we calculate taxes for the order based on the item. In that case, we get the information for the item from the existing cart, and then put back the information with the correct tax. As we are putting the item information from the cart, which is based on the images from Swell in the cart, how is it possible that we are putting back old images? The images we are getting are from existing cart from a Swell, so unless Swell is sending us back old images in cart call, there is no way we can be putting back old images. Where would that data come from? So it seems like Swell is sending us back old images for some reason in a cart call. Is it possible this is coming from some abandoned cart which had old images in the cart data? That's the only way I can see Swell sending back stale data.

**Reply by ericingram** [reply]
We haven't seen any cases where the API returns stale data, but even if it did, this could be avoided by not including the `items.product` object in the cart update. I can't say where it's coming from but it's definitely there in the logs and it has old image data in it.

There isn't a place in the API where product data is normally duplicated either, meaning carts/orders don't normally cache product data.

**Reply by osseonews** [reply]
Hmmm. I don't know. We will fix the tax calculation, but I don't see how that effects anything, as it is just duplicating whatever is in the cart and updating the taxes, and I'm pretty sure the product data is not even there, as you need expand the data to get product data in a cart call. We could not possibly update the product images otherwise, and we don't use expand in that api call. We don't put images anywhere or update the cart anywhere else, so I don't see how that is in the logs. Maybe your cart to orders function, example below, is doing something under the hood that is copying the items.product, b/c we certainly don't do that anywhere else. 

```
const order = (await this.client.post('/orders', {
        cart_id: cartId,
      })) as SwellOrder
```

**Reply by osseonews** [reply]
In any case, the dashboard routinely asks us if we want to save new data on a product, even if we didn't even update/change anything. Just opening up a product page in the dashboard, seems to trigger something which causes this, even if we didn't change a thing. My suspicion is that under the hood, when some API updates a product, there is some mismatch between what the dashboard has and what the API updated, if that makes sense, and then somehow the product gets updated to an earlier state.

I have to imagine that Swell caches the data somewhere for a product along with the associated images. Or am I wrong about that? Corrupted caches is not an uncommon problem.

**Reply by ericingram** [reply, marked as answer]
There is a bug related to the dashboard saying a page needs to be saved when you didn't change anything, we should prioritize that soon, but there isn't any data cache in the API or dashboard currently. I mean, there is cache logic, but it's been disabled for a long time as we need to chase down some race conditions before enabling it again.
